### ISSUE TITLE
The default value of showDialogKeyCombo (?) only works in certain keyboard layouts

### ISSUE THREAD

**Issue by rubenlg** [palantir/blueprint#4849]
#### Environment

- __Package version(s)__: @blueprintjs/core 3.47.0
- __Operating System__: Ubuntu 21.04
- __Browser name and version__: Electron 13.1.8 (reproduces on Chrome 92.0.4515.107 as well)

#### Code Sandbox

Link to a minimal repro: https://blueprintjs.com/docs/#core/hooks/use-hotkeys
(yes, that's your own documentation, it also reproduces there)

#### Steps to reproduce

1. Set your keyboard layout to Spanish ([example](https://upload.wikimedia.org/wikipedia/commons/thumb/7/74/KB_Spanish.svg/1280px-KB_Spanish.svg.png))
1. Focus the piano on the documentation.
1. Try to open the key combo dialog by typing `?`.

#### Actual behavior

Nothing. The `?` doesn't get registered. It also doesn't work as a regular "combo" in `HotkeyConfig`. 

Note that Electron and Chrome both catch that key if you use `document.body.onkeydown`. Also, on gmail, the `?` hotkey works well with my layout.

#### Expected behavior

The key combo dialog opens.

#### Possible solution

If I set my keyboard layout to English/US, then `?` works, both in `showDialogKeyCombo` as well as `HotkeyConfig`. Of course this is not a great workaround... I can't just ask all my users to change their keyboard layout to use my app. I don't know the blueprint codebase well enough to propose other alternatives. Happy to help with debugging if you point me in the right direction.

**Comment by rubenlg** [comment]
Found the root cause. [This code](https://github.com/palantir/blueprint/blob/1db3b449bebc31652115709750c9cce15714485b/packages/core/src/components/hotkeys/hotkeyParser.ts#L133) makes the incorrect assumption that `?` (and a bunch of other keys) always require pressing the `Shift` button, and creates a hardcoded association to the symbol that you would get in that button if you didn't press `Shift`. 

The reason that assumption is incorrect is that different countries use different keyboard layouts.

On that same file, `parseKeyCombo` claims that:

> For example, `@` is equivalent to `shift+2`.

That is wrong in Spanish, French and Belgian keyboards, to name a few.

**Comment by rubenlg** [comment]
At first I thought clearing the `ShiftKeys` object would solve the problem. However, that wasn't the only US-centric bug in the code. The code is also making the incorrect assumption that key codes can be mapped to constant printable characters. For example it assumes that key 219 always maps to `[`, but in a Spanish layout that key maps to `'`, and `?` when pressing `Shift`.

There is already an open issue (#4165) to migrate away of `which`, but the bug was open because it's deprecated, not because it's broken outside of the US.

This seems to require a significant refactor of `hotkeyParser.ts` and it's not trivial to fix because some of these incorrect dictionaries are exported, and most likely already depended upon by some US-based developers, so fixing this is going to involve a breaking change. Maybe something that could be shipped with blueprint 4.x, to justify the breaking change?

In the meantime, I think it would be worth pointing out in the documentation that hotkeys are broken for non-US keyboard layouts, so other developers don't go through the same process as I did.

**Comment by rubenlg** [comment]
Even if I switch the implementation to using `Event.key` and drop the code using `ShiftKeys`, I still have trouble with `?`, because I have to press `Shift` to produce that symbol, and `getKeyCombo` produces an object that won't match my `combo` spec since I don't want to include `shift` to make my code work in as many keyboard layouts as possible.

Here is an alternative idea for handling the `Shift` modifier:
- `Event.key` already encodes the _shift modifier info_ in many cases. For example, if you press shift+2, `Event.key` will contain the symbol above `2` on the layout of the user, not a `2`. For all these keys, there is no point in checking the shift modifier or matching it against the `combo` spec.
- A few keys don't change the value of `Event.key` when pressing `Shift`, and those could be handled separately. Examples are `Tab`, `F1`, `Esc`, etc. A whitelist for those would make sense.
- The alphabet is tricky. You can't just check `Event.key` because of caps-lock. In this case I'd recommend converting to lowercase and checking the shift modifier separately.

So the algorithm to process an event would be something like:
```javascript
if (isAlphabetic(ev.key)) {
  return {modifiers: modifiers, key: ev.key.toLocaleLowerCase() };
} else if (shiftHasNoInfluenceOn(ev.key)) {
  return {modifiers: modifiers, key: ev.key };
} else {
  // Drop the shift modifier.
  return {modifiers: modifiersWithoutShift, key: ev.key };
}
```
I realize that implementing `isAlphabetic` is not trivial. Besides `[a-zA-Z]`, you also need to handle `ñ`, `ç`, etc.

And when parsing a combo, the only special case is when the developer includes `shift` associated with a key that doesn't need it (e.g. `?`). In this case, `shift` could be dropped and a warning logged on the console.

Does this make sense?

**Comment by rubenlg** [comment]
Gentle ping.

**Comment by rubenlg** [comment]
Any thoughts on this?

**Comment by adidahiya** [comment]
@rubenlg thanks for your investigation into this issue and sharing your findings. I would like Blueprint hotkeys to work on non-US keyboard layouts, and you're right, it will require a bit of nontrivial work to do so. I would be open to the approach you described above, or perhaps borrowing something from other popular libraries which handle hotkeys. I can't get this in for the v4.0 release, but we should consider it for v5.0 coming later this year. Open to PRs...

**Comment by rubenlg** [comment]
Thank you for getting back to me. I lost the context on this already. If I find the time to map the code in my head again, I'll try to come up with a more detailed design and some PRs.

**Comment by adidahiya** [comment]
Circling back here since I recently made some changes to [`hotkeyParser.ts` for Blueprint v5 on the `next` branch](https://github.com/palantir/blueprint/blob/82e40024f73a2551087bd34c022461d971334223/packages/core/src/components/hotkeys/hotkeyParser.ts) - the bug still exists there, and we still make assumptions about the shift key being pressed to use the `?` key. Still open to bugfix PRs.

**Comment by rubenlg** [comment]
Thanks for the heads up! I haven't had the time to design a generic solution yet.
